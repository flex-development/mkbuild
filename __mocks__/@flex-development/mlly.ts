/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
