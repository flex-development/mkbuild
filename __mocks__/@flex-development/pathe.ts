/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
