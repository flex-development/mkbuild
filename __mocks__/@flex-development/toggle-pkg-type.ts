/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
