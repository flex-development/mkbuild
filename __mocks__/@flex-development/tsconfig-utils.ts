/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
