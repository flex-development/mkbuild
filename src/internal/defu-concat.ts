/**
 * @file Internal - defuConcat
 * @module mkbuild/internal/defuConcat
 */

import { isUndefined, type ObjectPlain } from '@flex-development/tutils'
import { mergeAndCompare as merge, type Merge } from 'merge-anything'

/**
 * Assigns default properties to the given `object`.
 *
 * Leftmost arguments have more priority when assigning defaults.
 *
 * Supports array concatenation.
 *
 * @template T - Destination object type
 * @template D - Source object type(s)
 *
 * @param {T} object - Destination object
 * @param {D} defaults - Object(s) used to assign default properties
 * @return {Merge<T, D> & T} `object` with defaults assigned
 */
function defuConcat<
  T extends ObjectPlain = ObjectPlain,
  D extends ObjectPlain[] = ObjectPlain[]
>(object: T, ...defaults: D): Merge<T, D> & T {
  /**
   * Determines how to assign a property.
   *
   * @param {unknown} val1 - Original value from {@linkcode object}
   * @param {unknown} val2 - Incoming value from object in {@linkcode defaults}
   * @return {unknown} Property value for {@linkcode object}
   */
  const compare = (val1: unknown, val2: unknown): unknown => {
    return Array.isArray(val1) && Array.isArray(val2)
      ? [...new Set([...val1, ...val2])]
      : isUndefined(val1)
      ? val2
      : val1
  }

  return merge<T, D>(compare, object, ...defaults) as Merge<T, D> & T
}

export default defuConcat
