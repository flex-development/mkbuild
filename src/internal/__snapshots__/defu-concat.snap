// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/defuConcat > should return object with defaults assigned 1`] = `
{
  "bundle": false,
  "clean": true,
  "cwd": ".",
  "entries": [],
  "ext": ".mjs",
  "format": "esm",
  "fs": {
    "lstat": [Function lstat],
    "mkdir": [Function mkdir],
    "readdir": [Function readdir],
    "readdirSync": [Function readdirSync],
    "rm": [Function rm],
    "stat": [Function stat],
    "unlink": [Function unlink],
    "writeFile": [Function writeFile],
  },
  "outdir": "dist",
  "pattern": "*.ts",
  "source": ".",
  "sourcemap": true,
  "sourcesContent": false,
  "target": [
    "node14.21.2",
    "esnext",
  ],
  "tsconfig": "tsconfig.build.json",
}
`;
